instrumentation-api-incubator/src/main/java/io/opentelemetry/instrumentation/api/incubator/semconv/messaging/MessagingConsumerMetrics.java

```diff
@@ -0,0 +1,114 @@
+ * metrics</a>.
+ */
+public final class MessagingConsumerMetrics implements OperationListener {
+  private static final double NANOS_PER_S = TimeUnit.SECONDS.toNanos(1);
+
+  // copied from MessagingIncubatingAttributes
+  private static final AttributeKey<Long> MESSAGING_BATCH_MESSAGE_COUNT =
+      AttributeKey.longKey("messaging.batch.message_count");
+  private static final ContextKey<MessagingConsumerMetrics.State> MESSAGING_CONSUMER_METRICS_STATE =
+      ContextKey.named("messaging-consumer-metrics-state");
+  private static final Logger logger = Logger.getLogger(MessagingConsumerMetrics.class.getName());
+
+  private final DoubleHistogram receiveDurationHistogram;
+  private final LongCounter receiveMessageCount;
+
+  private MessagingConsumerMetrics(Meter meter) {
+    DoubleHistogramBuilder durationBuilder =
+        meter
+            .histogramBuilder("messaging.receive.duration")
+            .setDescription("Measures the duration of receive operation.")
+            .setExplicitBucketBoundariesAdvice(MessagingMetricsAdvice.DURATION_SECONDS_BUCKETS)
+            .setUnit("s");
+    MessagingMetricsAdvice.applyReceiveDurationAdvice(durationBuilder);
+    receiveDurationHistogram = durationBuilder.build();
+
+    LongCounterBuilder longCounterBuilder =
+        meter
+            .counterBuilder("messaging.receive.messages")
+            .setDescription("Measures the number of received messages.")
+            .setUnit("{message}");
+    MessagingMetricsAdvice.applyReceiveMessagesAdvice(longCounterBuilder);
+    receiveMessageCount = longCounterBuilder.build();
+  }
+
+  public static OperationMetrics get() {
+    return OperationMetricsUtil.create("messaging consumer", MessagingConsumerMetrics::new);
+  }
+
+  @Override
+  @CanIgnoreReturnValue
+  public Context onStart(Context context, Attributes startAttributes, long startNanos) {
+    return context.with(
+        MESSAGING_CONSUMER_METRICS_STATE,
+        new AutoValue_MessagingConsumerMetrics_State(startAttributes, startNanos));
+  }
+
+  @Override
+  public void onEnd(Context context, Attributes endAttributes, long endNanos) {
+    MessagingConsumerMetrics.State state = context.get(MESSAGING_CONSUMER_METRICS_STATE);
+    if (state == null) {
+      logger.log(
+          FINE,
+          "No state present when ending context {0}. Cannot record consumer receive metrics.",
+          context);
+      return;
+    }
+
+    Attributes attributes = state.startAttributes().toBuilder().putAll(endAttributes).build();
+    receiveDurationHistogram.record(
+        (endNanos - state.startTimeNanos()) / NANOS_PER_S, attributes, context);
+
+    Long receiveMessagesCount = getReceiveMessagesCount(state.startAttributes(), endAttributes);
+    if (receiveMessagesCount != null) {
+      receiveMessageCount.add(receiveMessagesCount, attributes, context);
+    }
```

Comment:
I think this isn't quite correct. We should record the count of all messages but currently we are recording only the count of batch messages. Perhaps `getReceiveMessagesCount` should return `long` instead of `Long` and when `MESSAGING_BATCH_MESSAGE_COUNT` attribute is missing then return `1`.

Reply:
> This metric is [required](https://github.com/open-telemetry/semantic-conventions/blob/v1.26.0/docs/general/metric-requirement-level.md#required) for batch receive operations.
https://github.com/open-telemetry/semantic-conventions/blob/v1.26.0/docs/messaging/messaging-metrics.md#metric-messagingreceivemessages

This change will cause the `messaging.receive.messages` metric to appear in regular consumption as well, which does not align with the description in the spec.

Reply:
Imo you are misinterpreting the spec. It says that this metric is not required when messaging system does not support batch receives because the receive count can be obtained from a different metric. The metric description says `Measures the number of received messages.` which in my opinion means all messages not only messages that were received as a batch. Another question is since this metric is opt-in should `MessagingConsumerMetrics.get()` take a boolean argument that could be used to opt-in to this metric when instrumentation supports batch receives? @trask do you have any suggestions for this

Reply:
Thank you for your addition.

Reply:
let's align with the latest semconv which I think answer this question?